geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStore.scala

```diff
@@ -39,15 +40,28 @@ class KafkaDataStore(override val zookeepers: String,
 
   override def createTypeNames() = getNames.asScala.map(name => new NameImpl(getNamespaceURI, name.getLocalPart) : Name).asJava
 
+  case class FeatureSourceCacheKey(entry: ContentEntry, query: Query)
   private val featureSourceCache =
-    CacheBuilder.newBuilder().build[ContentEntry, ContentFeatureSource](
-      new CacheLoader[ContentEntry, ContentFeatureSource] {
-        override def load(entry: ContentEntry) = {
-          fsFactory(entry, kds)
+    CacheBuilder.newBuilder().build[FeatureSourceCacheKey, ContentFeatureSource](
+      new CacheLoader[FeatureSourceCacheKey, ContentFeatureSource] {
+        override def load(key: FeatureSourceCacheKey) = {
+          fsFactory(key.entry, kds)
         }
       })
 
-  override def createFeatureSource(entry: ContentEntry) = featureSourceCache.get(entry)
+
+  def getFeatureSource(typeName: String, filt: Filter): ContentFeatureSource = {
+    val entry = ensureEntry(new NameImpl(typeName))
+    val query = new Query(typeName, filt)
+    query.setStartIndex(0)
+    createFeatureSource(entry, query)
+  }
+
+  override def createFeatureSource(entry: ContentEntry) = createFeatureSource(entry, Query.ALL)
+  private def createFeatureSource(entry: ContentEntry, query: Query) = {
+    if(query.getStartIndex == null) query.setStartIndex(0)
```

Comment:
probably shouldn't use Query.ALL - query.getStartIndex will always be null, and then query.setStartIndex will throw an unsupported operation exception:
https://github.com/geotools/geotools/blob/master/modules/library/api/src/main/java/org/geotools/data/ALLQuery.java#L59
https://github.com/geotools/geotools/blob/master/modules/library/api/src/main/java/org/geotools/data/ALLQuery.java#L224